examples/text/plot_document_classification_20newsgroups.py

```diff
@@ -50,99 +37,198 @@
+    y_train, y_test = data_train.target, data_test.target
+
+    # Extracting features from the training data using a sparse vectorizer
+    t0 = time()
+    vectorizer = TfidfVectorizer(
+        sublinear_tf=True, max_df=0.5, min_df=5, stop_words="english"
+    )
+    X_train = vectorizer.fit_transform(data_train.data)
+    duration_train = time() - t0
+
+    # Extracting features from the test data using the same vectorizer
+    t0 = time()
+    X_test = vectorizer.transform(data_test.data)
+    duration_test = time() - t0
+
+    feature_names = vectorizer.get_feature_names_out()
+
+    if verbose:
+
+        # compute size of loaded data
+        data_train_size_mb = size_mb(data_train.data)
+        data_test_size_mb = size_mb(data_test.data)
+
+        print(
+            "%d documents - %0.3fMB (training set)"
+            % (len(data_train.data), data_train_size_mb)
+        )
+        print(
+            "%d documents - %0.3fMB (test set)"
+            % (len(data_test.data), data_test_size_mb)
+        )
+        print("%d categories" % len(target_names))
+        print(
+            "vectorize training done in %fs at %0.3fMB/s"
+            % (duration_train, data_train_size_mb / duration_train)
+        )
+        print("n_samples: %d, n_features: %d" % X_train.shape)
+        print(
+            "vectorize testing done in %fs at %0.3fMB/s"
+            % (duration_test, data_test_size_mb / duration_test)
+        )
+        print("n_samples: %d, n_features: %d" % X_test.shape)
+
+    return X_train, X_test, y_train, y_test, feature_names, target_names
 
-# %%
-# Vectorize the training and test data
-# -------------------------------------
-#
-# split a training set and a test set
-y_train, y_test = data_train.target, data_test.target
 
 # %%
-# Extracting features from the training data using a sparse vectorizer
-from time import time
+# Compare feature effects
+# -----------------------
+# First we train a model using the datasets with metadata. This will cause
+# overfitting
 
-from sklearn.feature_extraction.text import TfidfVectorizer
-from sklearn.feature_extraction.text import HashingVectorizer
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+from sklearn import metrics
+from sklearn.linear_model import RidgeClassifier
+from sklearn.metrics import ConfusionMatrixDisplay
 
-t0 = time()
+X_train, X_test, y_train, y_test, feature_names, target_names = load_dataset(
+    verbose=True
+)
 
-if USE_HASHING:
-    vectorizer = HashingVectorizer(
-        stop_words="english", alternate_sign=False, n_features=N_FEATURES
+clf = RidgeClassifier(tol=1e-2, solver="sparse_cg")
+clf.fit(X_train, y_train)
+pred = clf.predict(X_test)
+
+fig, ax = plt.subplots(figsize=(10, 5))
+ConfusionMatrixDisplay.from_predictions(y_test, pred, ax=ax)
+ax.xaxis.set_ticklabels(target_names)
+ax.yaxis.set_ticklabels(target_names)
+_ = ax.set_title("Confusion Matrix with headers for the\n" + str(clf).split("(")[0])
+
+
```

Comment:
Here again we can split: this way the figure for the confusion matrix appears right after the code that computes it and same for the features effects.